// @appium/base-driver/build/lib/express/idempotency.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleIdempotency = void 0;
const logger_1 = __importDefault(require("./logger"));
const lru_cache_1 = __importDefault(require("lru-cache"));
const lodash_1 = __importDefault(require("lodash"));
const events_1 = require("events");
const IDEMPOTENT_RESPONSES = new lru_cache_1.default({
    max: 64,
    ttl: 30 * 60 * 1000,
    updateAgeOnGet: true,
    updateAgeOnHas: true,
    dispose: (key, { responseStateListener }) => {
        responseStateListener?.removeAllListeners();
    }
});
const MONITORED_METHODS = ['POST', 'PATCH'];
const IDEMPOTENCY_KEY_HEADER = 'x-idempotency-key';
/**
 *
 * @param {string} key
 * @param {import('express').Request} req
 * @param {import('express').Response} res
 */
function cacheResponse(key, req, res) {
    if (!res.socket) {
        return;
    }
    const responseStateListener = new events_1.EventEmitter();
    IDEMPOTENT_RESPONSES.set(key, {
        method: req.method,
        path: req.path,
        response: null,
        responseStateListener,
    });
    const originalSocketWriter = res.socket.write.bind(res.socket);
    let response = '';
    const patchedWriter = (chunk, encoding, next) => {
        if (lodash_1.default.isString(chunk)) {
            response += chunk;
        }
        else if (lodash_1.default.isFunction(chunk.toString)) {
            response += chunk.toString(lodash_1.default.isString(encoding) ? encoding : undefined);
        }
        return originalSocketWriter(chunk, encoding, next);
    };
    res.socket.write = patchedWriter;
    let httpErrorMessage = null;
    let isResponseFullySent = false;
    res.once('error', (e) => { httpErrorMessage = e.message; });
    req.once('error', (e) => { httpErrorMessage = e.message; });
    res.once('finish', () => { isResponseFullySent = true; });
    res.once('close', () => {
        if (res.socket?.write === patchedWriter) {
            res.socket.write = originalSocketWriter;
        }
        if (!IDEMPOTENT_RESPONSES.has(key)) {
            logger_1.default.info(`Could not cache the response identified by '${key}'. ` +
                `Cache consistency has been damaged`);
        }
        else if (httpErrorMessage) {
            logger_1.default.info(`Could not cache the response identified by '${key}': ${httpErrorMessage}`);
            IDEMPOTENT_RESPONSES.delete(key);
        }
        else if (!isResponseFullySent) {
            logger_1.default.info(`Could not cache the response identified by '${key}', ` +
                `because it has not been completed`);
            logger_1.default.info('Does the client terminate connections too early?');
            IDEMPOTENT_RESPONSES.delete(key);
        }
        const value = IDEMPOTENT_RESPONSES.get(key);
        if (value) {
            value.response = Buffer.from(response, 'utf8');
            responseStateListener.emit('ready', value.response);
        }
        else {
            responseStateListener.emit('ready', null);
        }
    });
}
/**
 * @param {import('express').Request} req
 * @param {import('express').Response} res
 */
async function handleIdempotency(req, res, next) {
    const keyOrArr = req.headers[IDEMPOTENCY_KEY_HEADER];
    if (!keyOrArr) {
        return next();
    }
    const key = lodash_1.default.isArray(keyOrArr) ? keyOrArr[0] : keyOrArr;
    if (!MONITORED_METHODS.includes(req.method)) {
        // GET, DELETE, etc. requests are idempotent by default
        // there is no need to cache them
        return next();
    }
    logger_1.default.debug(`Request idempotency key: ${key}`);
    if (!IDEMPOTENT_RESPONSES.has(key)) {
        cacheResponse(key, req, res);
        return next();
    }
    const { method, path, response, responseStateListener, } = IDEMPOTENT_RESPONSES.get(key);
    if (req.method !== method || req.path !== path) {
        logger_1.default.warn(`Got two different requests with the same idempotency key '${key}'`);
        logger_1.default.warn('Is the client generating idempotency keys properly?');
        return next();
    }
    if (response) {
        logger_1.default.info(`The same request with the idempotency key '${key}' has been already processed`);
        logger_1.default.info(`Rerouting its response to the current request`);
        if (!res.socket?.writable) {
            return next();
        }
        res.socket.write(response.toString('utf8'));
    }
    else {
        logger_1.default.info(`The same request with the idempotency key '${key}' is being processed`);
        logger_1.default.info(`Waiting for the response to be rerouted to the current request`);
        responseStateListener.once('ready', async (/** @type {Buffer?} */ cachedResponse) => {
            if (!cachedResponse || !res.socket?.writable) {
                return next();
            }
            res.socket.write(cachedResponse.toString('utf8'));
        });
    }
}
exports.handleIdempotency = handleIdempotency;
